### ISSUE TITLE
numpy.random.randint(-2147483648, 2147483647) raises ValueError: low >= high (migrated from Trac #1690)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3241]
Original ticket http://projects.scipy.org/numpy/ticket/1690
Reported 2010-11-30 by trac user tkf, assigned to atmention:charris.

Following example raises "ValueError: low >= high".
It looks like it is due to overflow of "diff = hi - lo - 1" in RandomState.randint.

```
import numpy
intinfo = numpy.iinfo(int)
numpy.random.randint(-1, intinfo.max)  # raises nothing
numpy.random.randint(intinfo.min, intinfo.max)
## numpy.random.randint(-2, intinfo.max)  # raises ValueError too
```

**Comment by thouis** [comment]
Comment in Trac by trac user tkf, 2010-11-30

**Comment by thouis** [comment]
Attachment in Trac by trac user jpeel, 2011-01-04: [0001-BF-fix-for-randint-overflow.patch](http://projects.scipy.org/numpy/attachment/ticket/1690/0001-BF-fix-for-randint-overflow.patch)

**Comment by thouis** [comment]
Comment in Trac by trac user jpeel, 2011-01-04

It is clearly an overflow problem. I fixed the problem by:
1. checking explicitly if lo >= hi before raising the ValueError given above.
2. changing diff to an unsigned long, typecasting hi and lo to long longs before subtracting them, then typecasting the result to an unsigned long to be put in diff.

Everything is working quite well with these changes.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2011-01-24

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2011-01-24

Thanks for the patch. Fixed a bit differently in commit b7fbd3290e956dc1f9a8.

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2011-05-24
